go/storage/api/context.go

```diff
@@ -45,3 +52,47 @@ func NodePriorityHintFromContext(ctx context.Context) []signature.PublicKey {
 	nodes, _ := ctx.Value(contextKeyNodePriorityHint).([]signature.PublicKey)
 	return nodes
 }
+
+// WithNodeBlacklist sets a storage blacklist key for any storage requests using this context.
+// The blacklist is initially empty (i.e. all nodes are acceptable).
+func WithNodeBlacklist(ctx context.Context) context.Context {
+	return context.WithValue(ctx, contextKeyNodeBlacklist, map[signature.PublicKey]struct{}{})
+}
+
+// BlacklistAddNode adds a node to the blacklist associated with the context. If there's no
+// associated blacklist, the function does nothing.
+func BlacklistAddNode(ctx context.Context, node *node.Node) {
+	list, _ := ctx.Value(contextKeyNodeBlacklist).(map[signature.PublicKey]struct{})
+	if list != nil {
+		list[node.ID] = struct{}{}
+	}
+}
+
+// IsNodeBlacklistedInContext checks to see if the node is blacklisted in this context.
+// If the context doesn't have an associated blacklist, then no node is considered blacklisted.
+func IsNodeBlacklistedInContext(ctx context.Context, node *node.Node) bool {
+	val, ok := ctx.Value(contextKeyNodeBlacklist).(map[signature.PublicKey]struct{})
+	if !ok {
+		return false
+	}
+	_, ok = val[node.ID]
+	return ok
+}
+
+// NodeSelectionCallback is a callback used by the storage client to report connections used
+// for read requests.
+type NodeSelectionCallback = func(*node.Node)
+
+// WithNodeSelectionCallback sets a callback that will be called by the storage client on every read
+// request with the node that was eventually used to perform a successful request. If there was no
+// success, the callback isn't called.
+func WithNodeSelectionCallback(ctx context.Context, cb NodeSelectionCallback) context.Context {
```

Comment:
Could this be an interface instead? Maybe shared with the blacklist interface (maybe the interface needs a different name like `NodeManager` and the manager together with context accessors could be defined in `go/runtime/nodes` to be reusable in other contexts as well)?